### PR TITLE
feat: duplicate v4 APIs

### PR DESCRIPTION
**Pull request by jgiovaresco**
## Issue

https://gravitee.atlassian.net/browse/APIM-1183

## Description

Duplicate v4 APIs

## Additional context

---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-etjsnrcmky.chromatic.com)

### CODE REVIEW

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/resource/api/ApiResource.java [hunk 1]
```diff
@@ -277,6 +301,40 @@ public Response exportApiDefinition(@Context HttpHeaders headers, @PathParam("ap
...
+        GenericApiEntity duplicate;
+        var definitionVersion = currentEntity.getDefinitionVersion();
+        if (definitionVersion == DefinitionVersion.V4) {
+            duplicate =
+                duplicateApiService.duplicate(
+                    GraviteeContext.getExecutionContext(),
+                    (ApiEntity) currentEntity,
+                    DuplicateApiMapper.INSTANCE.mapToDuplicateOptions(duplicateOptions)
+                );
+        } else {
```
Comment (loriepisicchio) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/resource/api/ApiResource.java:345]: I think duplicating V1 API should not be allowed. Do we need to return a 400 bad request in that case ?

File: gravitee-apim-console-webui/src/management/api/general/details/api-general-info.component.html [hunk 2]
```diff
@@ -173,7 +173,7 @@
         *gioPermission="{ anyOf: ['api-definition-c'] }"
         mat-button
         class="details-card__actions_btn"
-        [disabled]="isKubernetesOrigin || api.definitionVersion === 'V4'"
+        [disabled]="isKubernetesOrigin"
```
Comment (loriepisicchio) [inline gravitee-apim-console-webui/src/management/api/general/details/api-general-info.component.html]: Not only related to this PR, but action should be disabled for V1 API.

**Review by loriepisicchio — APPROVED** [review]

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/resource/api/ApiResource.java [hunk 3]
```diff
@@ -277,6 +302,45 @@ public Response exportApiDefinition(@Context HttpHeaders headers, @PathParam("ap
             .build();
     }
 
+    @POST
+    @Path("/_duplicate")
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    @Permissions(
```
Comment (phiz71) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/resource/api/ApiResource.java]: To verify, but I think this list of permissions works like a "OR". IT means you need at least one of these two permissions. But in the openapi doc, it's written "AND".
I think you will have to change the doc or change the permissions check here.

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/mapper/DuplicateApiMapper.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+            return null;
+        }
+        var duplicateApiEntity = new DuplicateApiEntity();
+        duplicateApiEntity.setContextPath(duplicateApiOptions.getContextPath());
+        duplicateApiEntity.setVersion(duplicateApiOptions.getVersion());
+        duplicateApiEntity.setFilteredFields(
+            duplicateApiOptions.getFilteredFields().stream().map(DuplicateApiOptions.FilteredFieldsEnum::getValue).toList()
+        );
+
+        return duplicateApiEntity;
```
Comment (phiz71) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/mapper/DuplicateApiMapper.java]: Was it really required to write the full mapper by hand ?

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/resource/api/ApiResource.java [hunk 5]
```diff
@@ -277,6 +302,45 @@ public Response exportApiDefinition(@Context HttpHeaders headers, @PathParam("ap
...
+        }
+    )
+    public Response duplicateAPI(@PathParam("apiId") String apiId, @Valid @NotNull DuplicateApiOptions duplicateOptions) {
+        final GenericApiEntity currentEntity = getGenericApiEntityById(apiId, false);
+
+        GenericApiEntity duplicate;
+        var definitionVersion = currentEntity.getDefinitionVersion();
+
+        if (definitionVersion == DefinitionVersion.V1) {
+            throw new BadRequestException("Duplicating V1 API is not supported");
```
Comment (phiz71) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/resource/api/ApiResource.java]: you could use `return Response.status(Response.Status.BAD_REQUEST).entity(apiInvalid(apiId)).build();`

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/test/java/io/gravitee/rest/api/management/v2/rest/resource/api/ApiResource_DuplicateApiTest.java [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+        final Response response = rootTarget().request().post(Entity.json(aDuplicateApiOptions()));
+        assertThat(response.getStatus()).isEqualTo(FORBIDDEN_403);
+
+        var error = response.readEntity(Error.class);
+        assertThat(error.getHttpStatus()).isEqualTo(FORBIDDEN_403);
+        assertThat(error.getMessage()).isEqualTo("You do not have sufficient rights to access this resource");
+    }
+
+    @Test
+    public void should_return_412_when_duplicate_v1_api() {
```
Comment (phiz71) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/test/java/io/gravitee/rest/api/management/v2/rest/resource/api/ApiResource_DuplicateApiTest.java]: 400 not 412 ?

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/openapi/management-openapi-v2.yaml [hunk 7]
```diff
@@ -2272,6 +2301,30 @@ components:
...
+                filteredFields:
+                    type: array
+                    description: The list of API fields that can be excluded to create the new API.
+                    items:
+                        type: string
+                        enum:
+                            - groups
+                            - plans
+                            - members
+                            - pages
```
Comment (phiz71) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/openapi/management-openapi-v2.yaml]: All enums should be in uppercase in the openAPI doc

**Review by phiz71 — DISMISSED** [review]
"Request changes" is for the uppercase enum.

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/resource/api/ApiResource.java (hunk 3 above)
Comment (jgiovaresco) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/resource/api/ApiResource.java]: Good catch. I've copied/pasted from v1 rest api. I'm updating the check to ensure the AND

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/mapper/DuplicateApiMapper.java (hunk 4 above)
Comment (jgiovaresco) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/mapper/DuplicateApiMapper.java]: I've only kept to `mapToV2` to ensuire the lower case on filtered attributes

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/mapper/DuplicateApiMapper.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+import io.gravitee.rest.api.model.api.DuplicateApiEntity;
+import io.gravitee.rest.api.model.v4.api.DuplicateOptions;
+import org.mapstruct.Mapper;
+import org.mapstruct.factory.Mappers;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Mapper
+public interface DuplicateApiMapper {
+    Logger logger = LoggerFactory.getLogger(DuplicateApiMapper.class);
```
Comment (Okhelifi) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/mapper/DuplicateApiMapper.java]: Can we remove it ?

**Review by Okhelifi — APPROVED** [review]
Lgtm. Regarding the request changes on my side I'm ok to dismiss it as it's not an issue anymore. Good job 👍🏼

**Pull request merged**